test/fuzz/utils.hpp

```diff
@@ -0,0 +1,219 @@
+    get_next_input_data(&state->input, &state->context_num);
+    get_next_input_data(&state->input, &state->device_type_fuzz);
+    if (state->device_type_fuzz < 1 || state->device_type_fuzz > 7) {
+        return -1;
+    }
+    state->device_type = static_cast<ur_device_type_t>(state->device_type_fuzz);
+
+    return 0;
+}
+
+int get_next_api_call(FuzzerInput *input) {
+    uint8_t next_api_call;
+    if (get_next_input_data(input, &next_api_call) != 0) {
+        return -1;
+    }
+    return next_api_call % UR_MAX_FUZZER_API_CALL;
+}
+
+bool check_device_exists(const TestState *state) {
+    if (state->devices.empty() || state->device_num >= state->devices.size() ||
+        state->devices[0] == nullptr) {
+        return false;
+    }
+
+    return true;
+}
+
+bool check_context_exists(const TestState *state) {
+    if (state->contexts.empty() ||
+        state->context_num >= state->contexts.size() ||
+        state->contexts[0] == nullptr) {
+        return false;
+    }
+
+    return true;
+}
+
+bool check_pool_exists(
+    const std::map<ur_usm_pool_handle_t, std::vector<void *>> *map,
+    const uint8_t pool_num) {
+    if (pool_num >= map->size()) {
+        return false;
+    }
+
+    return true;
+}
+
+auto get_map_item(std::map<ur_usm_pool_handle_t, std::vector<void *>> *map,
+                  const uint8_t item_index) {
+    auto map_it = map->begin();
+    std::advance(map_it, item_index);
+
+    return map_it;
+}
+
+int get_alloc_pool(
+    TestState &state,
+    std::map<ur_usm_pool_handle_t, std::vector<void *>> &pool_map,
+    ur_usm_pool_handle_t pool, std::vector<void *> &allocs) {
+    uint8_t pool_num;
+
+    if (get_next_input_data(&state.input, &pool_num) != 0) {
+        return -1;
+    }
+    if (!check_pool_exists(&pool_map, pool_num)) {
+        return -1;
+    }
+
+    auto &[pool_tmp, allocs_tmp] = *get_map_item(&pool_map, pool_num);
+    pool = pool_tmp;
+    allocs = allocs_tmp;
+
+    return 0;
+}
+
+int load_kernel_source(std::vector<char> &binary_out) {
+    std::string source_path = KERNEL_IL_PATH;
+
+    std::ifstream source_file;
+    source_file.open(source_path,
+                     std::ios::binary | std::ios::in | std::ios::ate);
+    if (!source_file.is_open()) {
+        std::cerr << "Failed to open a kernel source file: " << source_path
+                  << std::endl;
+        return -1;
+    }
+
+    size_t source_size = static_cast<size_t>(source_file.tellg());
+    source_file.seekg(0, std::ios::beg);
+
+    std::vector<char> device_binary(source_size);
+    source_file.read(device_binary.data(), source_size);
+    if (!source_file) {
+        source_file.close();
+        std::cerr << "failed reading kernel source data from file: "
+                  << source_path << std::endl;
+        return -1;
+    }
+    source_file.close();
+
+    binary_out = std::vector<char>(std::move(device_binary));
+
+    return 0;
+}
+
+void cleanup(TestState *state) {
+    urLoaderConfigRelease(state->config);
```

Comment:
nit: Can we just wrap all the things that need to be releaed/freed/closed in unique_ptrs instead of implementing this function?

Reply:
Done